Coalesce TLS accesses in RecordFunction constructor

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44970 Coalesce TLS accesses in RecordFunction constructor**

Summary:
Reducing number of TLS accesses from RecordFunction

Test Plan:
record_function_benchmark
callgrind: https://gist.github.com/ilia-cher/0fc6befeeb7628409fa0c0a5b94385a2

Differential Revision: [D23791165](https://our.internmc.facebook.com/intern/diff/D23791165)

Commit:
[wip] Coalesce TLS accessed in RecordFunction constructor

Summary:
Right now, when RecordFunction is not active (usual case),
we do two TLS accesses (check for thread local callbacks, and check for
thread local boolean).
Experimenting with reducing number of TLS accesses in RecordFunction
constructor.

Test Plan:
record_function_benchmark

Commit:
Update on "[wip] Coalesce TLS accesses in RecordFunction constructor"

[wip] Coalesce TLS accessed in RecordFunction constructor

Summary:
Right now, when RecordFunction is not active (usual case),
we do two TLS accesses (check for thread local callbacks, and check for
thread local boolean).
Experimenting with reducing number of TLS accesses in RecordFunction
constructor.

Test Plan:
record_function_benchmark

gh-metadata: pytorch pytorch 44970 gh/ilia-cher/41/head

Commit:
Update on "[wip] Coalesce TLS accesses in RecordFunction constructor"

[wip] Coalesce TLS accessed in RecordFunction constructor

Summary:
Right now, when RecordFunction is not active (usual case),
we do two TLS accesses (check for thread local callbacks, and check for
thread local boolean).
Experimenting with reducing number of TLS accesses in RecordFunction
constructor.

Test Plan:
record_function_benchmark

gh-metadata: pytorch pytorch 44970 gh/ilia-cher/41/head

Commit:
Update on "[wip] Coalesce TLS accesses in RecordFunction constructor"

[wip] Coalesce TLS accessed in RecordFunction constructor

Summary:
Right now, when RecordFunction is not active (usual case),
we do two TLS accesses (check for thread local callbacks, and check for
thread local boolean).
Experimenting with reducing number of TLS accesses in RecordFunction
constructor.

Test Plan:
record_function_benchmark

gh-metadata: pytorch pytorch 44970 gh/ilia-cher/41/head

dzhulgakov reviewed (approved):
The changes themselves look good. For completeness - can you share calgrind instruction numbers and also run benchmark in a less noisy environment? It's interesting to see whether it makes any difference.

Commit:
Update on "[wip] Coalesce TLS accesses in RecordFunction constructor"


Summary:
Right now, when RecordFunction is not active (usual case),
we do two TLS accesses (check for thread local callbacks, and check for
thread local boolean).
Experimenting with reducing number of TLS accesses in RecordFunction
constructor.

Test Plan:
record_function_benchmark

https://gist.github.com/ilia-cher/8b6cfa26d9d84cfa3435eb384fa93cf6

Differential Revision: [D23791165](https://our.internmc.facebook.com/intern/diff/D23791165)

[ghstack-poisoned]

ilia-cher commented:
Dmytro (@dzhulgakov),  the TLS accesses were primarily from ThreadLocalState (updated this PR)
with an update, we get callgrind results: https://gist.github.com/ilia-cher/0fc6befeeb7628409fa0c0a5b94385a2
7841674 / 7651084 = 1.025 -> ~2.5% reduction in instructions
__tls_get_addr: 279400 / 250800 = 1.11 -> ~11% reduction in __tls_get_addr
(__tls_get_addr_slow: 127000 / 114000 = 1.11)

ilia-cher commented:
(@ezyang, @bhosmer, @ngimel)
